### ISSUE TITLE
UB when reading ImmutableImage with General layout

### ISSUE THREAD

**Issue by faulesocke** [vulkano-rs/vulkano#1447]
I get random behaviour (so likely this is UB) when reading a `ImmutableImage` with layout `General` using `AutoCommandBufferBuilder::blit_image`. The minimum example below reproduces this error:

```rust
use vulkano::{
    buffer::{BufferUsage, CpuAccessibleBuffer},
    command_buffer::{AutoCommandBufferBuilder, CommandBuffer},
    device::{Device, DeviceExtensions},
    format::Format,
    image::{Dimensions, ImageLayout, ImageUsage, ImmutableImage},
    instance::{Instance, InstanceExtensions, PhysicalDevice},
    sampler::Filter,
    sync::GpuFuture,
};

fn main() {
    let instance = Instance::new(None, &InstanceExtensions::none(), None).unwrap();
    let physical = PhysicalDevice::enumerate(&instance).next().unwrap();
    println!(
        "Using device: {} (type: {:?})",
        physical.name(),
        physical.ty()
    );

    let queue_family = physical
        .queue_families()
        .find(|&q| q.supports_graphics())
        .unwrap();

    let (device, mut queues) = Device::new(
        physical,
        physical.supported_features(),
        &DeviceExtensions::none(),
        [(queue_family, 0.5)].iter().cloned(),
    )
    .unwrap();
    let queue = queues.next().unwrap();

    let dimensions = Dimensions::Dim2d {
        width: 128,
        height: 128,
    };
    let format = Format::R8G8B8A8Srgb;
    let usage = ImageUsage {
        transfer_source: true,
        transfer_destination: true,
        ..ImageUsage::none()
    };
    let tex1 = {
        let data = vec![0; 128 * 128 * 4];

        // NOTE THIS
        let layout = ImageLayout::General;

        let (tex, init) = ImmutableImage::uninitialized(
            device.clone(),
            dimensions,
            format,
            1,
            usage,
            layout,
            device.active_queue_families(),
        )
        .unwrap();

        let buffer = CpuAccessibleBuffer::from_iter(
            device.clone(),
            BufferUsage::transfer_source(),
            false,
            data.into_iter(),
        )
        .unwrap();

        let mut cbb =
            AutoCommandBufferBuilder::primary_one_time_submit(device.clone(), queue_family)
                .unwrap();
        cbb.copy_buffer_to_image(buffer, init).unwrap();

        let cb = cbb.build().unwrap();
        cb.execute(queue.clone())
            .unwrap()
            .then_signal_fence_and_flush()
            .unwrap()
            .wait(None)
            .unwrap();

        tex
    };

    let tex2 = {
        let layout = ImageLayout::General;
        let (tex, init) = ImmutableImage::uninitialized(
            device.clone(),
            dimensions,
            format,
            1,
            usage,
            layout,
            device.active_queue_families(),
        )
        .unwrap();

        let mut cbb =
            AutoCommandBufferBuilder::primary_one_time_submit(device.clone(), queue_family)
                .unwrap();
        cbb.blit_image(
            tex1.clone(),
            [0, 0, 0],
            [
                dimensions.width() as _,
                dimensions.height() as _,
                dimensions.depth() as _,
            ],
            0, // source base array layer
            0, // source mip level
            init,
            [0, 0, 0],
            [
                dimensions.width() as _,
                dimensions.height() as _,
                dimensions.depth() as _,
            ],
            0, // dest base array layer
            0, // dest mip level
            1, // layer count
            Filter::Linear,
        )
        .unwrap();

        let cb = cbb.build().unwrap();
        cb.execute(queue.clone())
            .unwrap()
            .then_signal_fence_and_flush()
            .unwrap()
            .wait(None)
            .unwrap();

        tex
    };

    dbg!(tex1.dimensions());
    dbg!(tex2.dimensions());
}
```

In all observed cases the program panics instead of printing the two debug outputs. I observed two cases where the panic occours: In the first case the second command buffer execute returns an `Err` variant and then the `unwrap` causes a panic:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AccessError { error: ExclusiveDenied, command_name: "vkCmdBlitImage", command_param: "source", command_offset: 0 }', src/main.rs:128:14
stack backtrace:
[... just stdlib stuff ...]
   4: vulkanobugs::main
             at ./src/main.rs:127:9
[... just stdlib stuff ...]
```

In the second case an `assert!` within vulkano itself fails:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `None`,
 right: `Some(General)`', /home/socke/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/image/immutable.rs:536:9
stack backtrace:
[... just stdlib stuff ...]
   2: <vulkano::image::immutable::ImmutableImageInitialization<F,A> as vulkano::image::traits::ImageAccess>::unlock
             at /home/X/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/image/immutable.rs:536:9
   3: vulkano::command_buffer::synced::base::SyncCommandBuffer<P>::lock_submit
             at /home/X/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/command_buffer/synced/base.rs:1218:29
   4: <vulkano::command_buffer::auto::AutoCommandBuffer<P> as vulkano::command_buffer::traits::CommandBuffer>::lock_submit
             at /home/X/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/command_buffer/auto.rs:1761:25
   5: vulkano::command_buffer::traits::CommandBuffer::execute_after
             at /home/X/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/command_buffer/traits.rs:141:9
   6: vulkano::command_buffer::traits::CommandBuffer::execute
             at /home/X/.cargo/registry/src/github.com-1ecc6299db9ec823/vulkano-0.19.0/src/command_buffer/traits.rs:97:9
   7: vulkanobugs::main
             at ./src/main.rs:127:9
[... just stdlib stuff ...]
```

I haven't fully investigated the problem yet but one interesting detail should be mentioned: When changing the layout of the first image from `General` to `TransferSrcOptimal` the problem disappears. So probably some layout transition happening around the blit command is not correct. The problem also occours when using `copy_image` instead of `blit_image`.

*   Version of vulkano: 0.19
*   OS: Linux
*   GPU (the selected PhysicalDevice): Intel (probably not relevant)

**Comment by Gosuo** [comment]
@faulesocke This has probably something to do with this: [code](https://github.com/vulkano-rs/vulkano/blob/master/vulkano/src/command_buffer/synced/base.rs#L1218-L1224)?

It unlocks all images with `None`, I guess that is not correct. [Here](https://github.com/vulkano-rs/vulkano/blob/master/vulkano/src/command_buffer/synced/base.rs#L1262-L1271) it unlocks the images with the `final_layout` if that differs from the `initial_layout`.

Unfortunately, I'm not too familiar with image operations in Vulkan, so I'm not sure what the intended operation would be. Would it need to be unlocked with the `final_layout` or the `initial_layout`? I guess the latter as the operation wasn't completed at that point.

**Comment by faulesocke** [comment]
Locking is entirely vulkano-specific. From what I have seen I guess the image needs to be unlocked with the desired `final_layout`.

Layouts describe the way the GPU stores data for images inside memory. The vulkan cookbook describes this the following way:

> For example, sampling data from images inside shaders may need them to be cached in such a way that neighbor texels are also neighbors in memory. But, writing data to images may be performed faster when a memory is laid out linearly.

So before doing any operation on an image you usually have to transition it to a given layout. Transitioning happens with inserting memory barriers into command buffers and vulkano usually does this work automatically for us. There is also a `General` layout which works for all operations but might be slower. When specifying an image layout in vulkano what you really do is telling vulkano that each command buffer should leave the image in this layout after it's done. Of course in the case of a `General` layout transitions are not necessary and vulkano probably tries to detect this already.

I will dig a bit through the code to see what really happens and whether operations are sound the way they are.

**Comment by Eliah-Lakhin** [comment]
@faulesocke Can you confirm if this issue resolved after #1449 merge?

**Comment by Gosuo** [comment]
I don't think that #1449 fixes the actual issue, just the assertion panic that is happening as well.
